src/toolchain.rs

```diff
@@ -374,6 +374,37 @@ impl<'a> Toolchain<'a> {
         Ok(None)
     }
 
+    fn maybe_do_rust_analyzer_fallback(&self, binary: &str) -> anyhow::Result<Option<Command>> {
+        if binary != "rust-analyzer" && binary != "rust-analyzer.exe"
+            || self.binary_file("rust-analyzer").exists()
+        {
+            return Ok(None);
+        }
+
+        let proc = self.cfg.process;
+        let Some(path) = proc.var_os("PATH") else {
+            return Ok(None);
+        };
+
+        let me = env::current_exe()?;
+
+        // Try to find the first `rust-analyzer` under the `$PATH` that is both
+        // an existing file and not the same file as `me`, i.e. not a rustup proxy.
+        Ok(env::split_paths(&path).find_map(|mut p| {
+            p.push(binary);
```

Comment:
We might want to check the executable bit of this path. And maybe show the first nonexecutable file path when nothing found.

Reply:
I'm not too sure about that. If something is in PATH the expectation is that it's executable. It's up to the OS to determine if it actually is or not and I don't think it's useful to second guess it.

Reply:
I agree this should look more like `which` in that only executables can be considered candidates. However I'm not very sure about the second half...

Reply:
And most shells have the behavior, as well as [execvp](https://linux.die.net/man/3/execvp) on Linux.

The other way is falling back to call `rust-analyzer` directly and let OS do whatever it wants, so we don't need to mimic the PATH lookup logic.

Reply:
@ChrisDenton @weihanglo Thanks for the detailed discussion touching both sides of the argument. I just read the relevant [`which-rs`' code](https://github.com/harryfei/which-rs/blob/eef199824a0cf1596e8afbe9e7a5e6a793486cad/src/checker.rs#L14-L44) on this and what it's doing is basically checking `x` on Unix exclusively. I guess that'd be a reasonable middle ground for this particular case (and stuff like a TOCTOU error is not really important here I believe)...

Reply:
> Btw, I just tried in bash and it just runs a script without the executable bit (well, it fails, but it tries) and does not read the rest of the PATH.

I just tried the same on my Mac and I believe that's not the case at least on my machine: a simple `touch rust-analyzer` can change the interpretation of calling that command depending on `chmod +x` and `chmod -x`.

Reply:
Seems like there's a difference between bash and which:

```
$ which biscuits
/home/chris/.cargo/bin/biscuits
$ biscuits
-bash: /home/chris/temp/biscuits: Permission denied
```

Reply:
@ChrisDenton 
Did you have another biscuits executable with a lower precedence in your PATH?

That was not my case with bash 5.2.37 on both macOS 15.4.1 and Linux 5.10.235. Not sure what was different 🤔

Reply:
Oh I see. `/home/chris/.cargo/bin/biscuits` is the executable

Reply:
I'm not quite sure, but anyway this is what I've tried, and it seems to be coherent with what `which` gives:

```console
> printf '#!/bin/sh\necho Hello World' > ~/.moon/bin/rust-analyzer
> chmod +x ~/.moon/bin/rust-analyzer
> rust-analyzer
Hello World
> chmod -x ~/.moon/bin/rust-analyzer
> rust-analyzer
error: Unknown binary 'rust-analyzer' in official toolchain 'stable-aarch64-apple-darwin'.
```

Maybe that's because I'm using `fish`?

Reply:
I've updated my solution to align with [`which-rs`](https://github.com/harryfei/which-rs/blob/eef199824a0cf1596e8afbe9e7a5e6a793486cad/src/checker.rs#L14-L44) as mentioned above.